File: Simplenote/SPWindow.m
#import "SPWindow.h"
#import "Simplenote-Swift.h"


#pragma mark - SPWindow

@implementation SPWindow

- (void)dealloc
{
    [self stopListeningToNotifications];
}

- (instancetype)initWithContentRect:(NSRect)contentRect styleMask:(NSWindowStyleMask)styleMask backing:(NSBackingStoreType)bufferingType defer:(BOOL)flag
{
    if (self = [super initWithContentRect:contentRect styleMask:styleMask backing:bufferingType defer:flag]) {
        [self startListeningToNotifications];
        [self reloadAppearance];
    }

    return self;
}

- (instancetype)initWithContentRect:(NSRect)contentRect styleMask:(NSWindowStyleMask)styleMask backing:(NSBackingStoreType)bufferingType defer:(BOOL)flag screen:(NSScreen *)screen
{
    if ((self = [super initWithContentRect:contentRect styleMask:styleMask backing:bufferingType defer:flag screen:screen])) {
        [self startListeningToNotifications];
        [self reloadAppearance];
    }

    return self;
}


#pragma mark - Initialization Helpers

- (void)startListeningToNotifications
{
    NSNotificationCenter *nc = [NSNotificationCenter defaultCenter];
    [nc addObserver:self selector:@selector(reloadAppearance) name:ThemeDidChangeNotification object:nil];
}

- (void)stopListeningToNotifications
{
    [[NSNotificationCenter defaultCenter] removeObserver:self];
}

- (void)reloadAppearance
{
    if (@available(macOS 10.14, *)) {
        if ([SPUserInterface isSystemThemeSelected]) {
            self.appearance = nil;
            return;
        }

        NSAppearanceName name = [SPUserInterface isDark] ? NSAppearanceNameVibrantDark : NSAppearanceNameVibrantLight;
        self.appearance = [NSAppearance appearanceNamed:name];
    }
}

@end

File: Simplenote/SPWindow.swift
import Foundation
import AppKit


// MARK: - Simplenote Window
//
class SPWindow: NSWindow {

    deinit {
        stopListeningToNotifications()
    }

    override init(contentRect: NSRect, styleMask style: NSWindow.StyleMask, backing backingStoreType: NSWindow.BackingStoreType, defer flag: Bool) {
        super.init(contentRect: contentRect, styleMask: style, backing: backingStoreType, defer: flag)
        startListeningToNotifications()
        refreshStyle()
    }
}


// MARK: - Initialization
//
private extension SPWindow {

    func startListeningToNotifications() {
        NotificationCenter.default.addObserver(self, selector: #selector(refreshStyle), name: .ThemeDidChange, object: nil)
    }

    func stopListeningToNotifications() {
        NotificationCenter.default.removeObserver(self)
    }
}


// MARK: - Notification Handlers
//
private extension SPWindow {

    @objc
    func refreshStyle() {
        guard #available(macOS 10.14, *) else {
            return
        }

        if SPUserInterface.isSystemThemeSelected {
            appearance = nil
            return
        }

        appearance = NSAppearance.simplenoteAppearance
    }
}
